#
# This is a Shiny web application. You can run the application by clicking
# the 'Run App' button above.
#
# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#

library(shiny)
library(shinydashboard)

# Define UI for application that draws a histogram
ui <- fluidPage(
   
  dashboardPage(skin = "green",
                dashboardHeader(title = "Title"),
                dashboardSidebar(
                  sidebarMenu(
                    id = "inTabset",
                    menuItem("Tutorial", tabName = "help"),
                    menuItem("Acknowledgments", tabName = "acknowledgments"),
                    menuItem("FAQ", tabName = "faq")
                  )
                ),
                dashboardBody(
                  tabItems(
                    tabItem(tabName = "help",
                            box(width = 3, height = '90vh', title = "Summary", solidHeader = TRUE, status = "primary",
                                div(style='height: 80vh; overflow-y: scroll',
                                    tags$li(tags$a(href = "#1", "Get started"),
                                            tags$ul(tags$li(tags$a(href = "#1_1", "Enzyme query"),
                                                            tags$ul(tags$li(tags$a(href = "#1_1_1", "EC Number")),
                                                                    tags$li(tags$a(href = "#1_1_2", "Enzyme name"))
                                                            ))
                                            ),
                                            tags$ul(tags$li(tags$a(href = "#1_2", "Before the query"),
                                                            tags$ul(tags$li(tags$a(href = "#1_2_1", "Add enzymes to the query")),
                                                                    tags$li(tags$a(href = "#1_2_2", "Erase the query"))
                                                            ))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#2", "Queries from the organism table"),
                                            tags$ul(tags$li(tags$a(href = "#2_1", "Get PDB")),
                                                    tags$li(tags$a(href = "#2_2", "Get amino acids sequence")),
                                                    tags$li(tags$a(href = "#2_3", "Get functional Parameters"))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#3", "Functional Parameters Query"),
                                            tags$ul(tags$li(tags$a(href = "#3_1", "Selecting parameters")),
                                                    tags$li(tags$a(href = "#3_2", "Other options of query"),
                                                            tags$ul(tags$li(tags$a(href = "#3_2_1", "Selecting by phylogeny")),
                                                                    tags$li(tags$a(href = "#3_2_2", "Add enzymes")),
                                                                    tags$li(tags$a(href = "#3_2_3", "Filter by Uniprot code"))
                                                            )
                                                    ),
                                                    tags$li(tags$a(href = "#3_3", "The Parameter table")),
                                                    tags$li(tags$a(href = "#3_4", "Filters")),
                                                    tags$li(tags$a(href = "#3_5", "Download table"))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#4", "Summary table and quick access"),
                                            tags$ul(tags$li(tags$a(href = "#4_1", "Summary table and where to find it")),
                                                    tags$li(tags$a(href = "#4_2", "Quick access toolbar")),
                                                    tags$li(tags$a(href = "#4_3", "Dashboard Menu"))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#5", "Visualize"),
                                            tags$ul(tags$li(tags$a(href = "#5_1", "Parameter found as a bar plot")),
                                                    tags$li(tags$a(href = "#5_2", "Distribution as boxplot")),
                                                    tags$li(tags$a(href = "#5_3", "Correlation"),
                                                            tags$ul(tags$li(tags$a(href = "#5_3_1", "Correlation as a heatmap matrix")),
                                                                    tags$li(tags$a(href = "#5_3_2", "Correlation as a paired scatter plot"))
                                                            ))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#6", "Cluster"),
                                            tags$ul(tags$li(tags$a(href = "#6_1", "K-means")),
                                                    tags$li(tags$a(href = "#6_2", "DBSCAN"))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#7", "External tools"),
                                            tags$ul(tags$li(tags$a(href = "#7_1", "Protein Prediction")),
                                                    tags$li(tags$a(href = "#7_2", "Sequence Alignment"))
                                            )
                                    ),
                                    tags$li(tags$a(href = "#8", "Saved tables")
                                    ),
                                    tags$li(tags$a(href = "#9", "Errors"),
                                            tags$ul(tags$li(tags$a(href = "#9_1", "Avoidable errors"),
                                                            tags$ul(tags$li(tags$a(href = "#9_1_1", "Enter Brenda User")),
                                                                    tags$li(tags$a(href = "#9_1_2", "Wrong Password")),
                                                                    tags$li(tags$a(href = "#9_1_3", "Not a Brenda Account")),
                                                                    tags$li(tags$a(href = "#9_1_4", "Make an enzyme query")),
                                                                    tags$li(tags$a(href = "#9_1_5", "Cannot add")),
                                                                    tags$li(tags$a(href = "#9_1_6", "Make a functional parameters query")),
                                                                    tags$li(tags$a(href = "#9_1_7", "A wrong number of parameter selected"))
                                                            )),
                                                    tags$li(tags$a(href = "#9_2", "Unavoidable errors"),
                                                            tags$ul(tags$li(tags$a(href = "#9_2_1", "Slower Algorithm")),
                                                                    tags$li(tags$a(href = "#9_2_2", "Unexpected error on search")),
                                                                    tags$li(tags$a(href = "#9_2_3", "The parameters appear selected, but the query doesn't show anything")),
                                                                    tags$li(tags$a(href = "#9_2_4", "Reduce the obtained rows")),
                                                                    tags$li(tags$a(href = "#9_2_5", "Empty tables")),
                                                                    tags$li(tags$a(href = "#9_2_6", "Suggestions"))
                                                            ))
                                            )
                                    )
                                )
                            ),
                            box(width = 9, height = '90vh', title = h1(icon("graduation-cap"), "Tutorial"), solidHeader = TRUE, status = "primary",
                                div(style='height:70vh; overflow-y: scroll',
                                    div(id = "1", class = "section level1",
                                        h2("Get started"),
                                        p("The enzymes are a complex group of proteins, which are essential for most biological processes in ",
                                          "living systems. This characteristic gives them great potential in biotechnological applications."),
                                        p("Currently, the most whole enzyme database worldwide is  BRaunschweig ENzyme DAtabase (BRENDA), ",
                                          "which has 104,000 enzymes names with their functional parameters and other properties. BRENDA allows ",
                                          "access to this information. However, the interface is complex and the data is hard to aggregate."),
                                        p("We propose Brenda Easy Search Tool (BEST) a versatile web tool that enables querying enzymes and proteins ",
                                          "in the BRENDA database fast and efficiently. This tool allows you to obtain functional parameters and ",
                                          "molecular properties from your enzyme query in a user-friendly format. This tool was developed in Shiny R."),
                                        p("To search for data in BRENDA, we need your Brenda account (",
                                          actionLink("faqhelp", "why?"), ").",
                                          "If you do not have one, you can visit ",
                                          tags$a("Brenda Registration.", href="https://www.brenda-enzymes.org/register.php", target = "_blank"),
                                          "Once you have your BRENDA's account, we can access to the whole collection of enzymes and obtained the ones ",
                                          "you need, and the numerical parameters you request.")),
                                    div(id = "1_1", class = "section level2",
                                        h3("Enzyme query"),
                                        p("To get the enzymes you need, we offer two different kinds of options. These inputs are in ",
                                          "the ",
                                          actionLink("enzymeHelp", "select enzyme"),
                                          "section.")),
                                    div(id = "1_1_1", class = "section level3",
                                        h4("EC Number"),
                                        p("Enzymes are classified by the ",
                                          tags$a("IUBMB", href = "https://iubmb.org/", target = "_blank"),
                                          " according to the reaction they catalyze. This corresponds to the",
                                          "EC Number, which is a four number code, with the first one indicating the subclass ",
                                          "such as oxidoreductases, transferases, hydrolases, lyases, isomerases, ligases and translocases. ",
                                          "The second number indicates the functional group of reaction, the third indicates a specific ",
                                          "type of molecule and the fourth the last",
                                          "level and the reaction itself. More info: ",
                                          tags$a("EC number.", href = "https://www.qmul.ac.uk/sbcs/iubmb/", target = "_blank"),
                                          br(),
                                          "If you know the EC number of your enzyme of interest, indicate it in the input",
                                          "shown in the next image: "),
                                        img(src = 'Tutorial(1).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "1_1_2", class = "section level3",
                                        h4("Enzyme name"),
                                        p("Enzymes have a lot of names that differ depending on the gen codifying it, the reaction, ",
                                          "historical reasons and so on. Brenda groups them according to its EC number, however, they are associated",
                                          "with the multiple names in which they are called in literature. This category is shown on the ",
                                          tags$a("Brenda web page", href = "https://www.brenda-enzymes.org/", target = "_blank"),
                                          "with the name 'Synonyms'."),
                                        p("Due to the length of the list, that contains every synonym of every type of enzyme on the database, these names ",
                                          "are separated in subclasses. Then, if you know just a particular name of your enzyme, first you have to",
                                          " know in which of these categories it is: "),
                                        img(src = 'Tutorial(2).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("For the next step, you have to select the functional group that is involved in the catalyzed ",
                                          "reaction, such as: "),
                                        img(src = 'Tutorial(3).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("In the following space, you can type the name of the enzyme you know. You can search for the list ",
                                          "below, remembering that it is a long list. This selector autocompletes your input and shows the possible ",
                                          "names that are registered in the Brenda database, select the one that fits most with the name you know."),
                                        img(src = 'Tutorial(4).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("After you press the search button, a short list of the possible type of enzyme is going to be shown (it can be ",
                                          "just one). The selector shows the EC number and Brenda's recommended name. Select the one you are looking ",
                                          "for, press 'go', and go on with the numerical parameter selector."),
                                        img(src = 'Tutorial(5).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "1_2", class = "seccion level2",
                                        h3("Before the query"),
                                        p("At this point, a table showing all the organism that has a protein with the same catalytic function defined by ",
                                          "EC number, available in Brenda is generated. This table contains the next columns that depend on the enzyme type: ",
                                          "EC number, systematic name (this could be empty) and recommended name; and the next columns are related to the organism: ",
                                          "UniProt, organism (scientific name) and commentary. This last one is hidden by default, but can be shown selecting the ",
                                          "\"show the commentary column\" checkbox."),
                                        img(src = 'Tutorial(6).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("After the ", tags$a("parameter query is done,", href = "#3"), " the protein table contains a \"Literature\" column that ",
                                          "can be shown clicking the checkbox next to the \"Show commentary columns\"."),
                                        p("The way to proceed after this point is described in ", tags$a("\"Queries from the organism table\" section.", href = "#2"),
                                          " In this section, the optional proceeds to have a complex query are described.")
                                        ),
                                    div(id = "1_2_1", class = "seccion level3",
                                        h4("Add Enzymes to the query"),
                                        p("If more than just one enzyme type is needed, you can add another EC number to the query. This process can be done virtually ",
                                          "indefinitely. To add EC numbers to your organism table (or protein table) look for the \"Add proteins\" button in the sidebar ",
                                          "panel."),
                                        img(src = 'Tutorial(7).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("After clicking it, you are going to be back at the",
                                          actionLink("enzymeHelp2", "enzyme section"),
                                          ", make sure the welcome panel in the top of the page ",
                                          "(just below the toolbar) says \"",icon("exclamation-triangle"),"The enzyme you enter will be added to the search\"",
                                          "if this is not the case, please go back to the Protein table section (by dashboard menu or by the quick access button in ",
                                          "toolbar shown below) and press the \"Add proteins\" button again."),
                                        img(src = 'Tutorial(8).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("If all is set, add enzymes to the query ", tags$a("the same way the first ones were entered.", href = "#1_1"))
                                        ),
                                    div(id = "1_2_2", class = "seccion level3",
                                        h4("Erase the query"),
                                        p("If you made a mistake and the EC number enter to the query is not the one you want to. You can erase this query and ",
                                          "look for another one by clicking the \"Go back and search another Enzyme\" in the top/right corner."),
                                        img(src = 'Tutorial(9).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("In this case, the welcome panel should show: "),
                                        img(src = 'Tutorial(10).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("Sadly, if you had more than one enzymes in the query, all of them are going to be erased.",
                                          div(style = "display: inline-block;font-weight: bold;",
                                              "There is no way to just erase the last one.")),
                                        p("Log out has the same effect. This is going to erase everything generated on the site. But, this has the advantage that ",
                                          "specific tables are going to be", tags$a("saved", href = "#8")),
                                        img(src = 'Tutorial(11).jpg',
                                            width = '350px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "2", class = "seccion level1",
                                        h2("Queries from the organism table"),
                                        p("After the organism table ",
                                          "(", actionLink("proteinHelp", "Protein Table"),
                                          ") is generated, it can be used to search for ", tags$a("numerical parameters,", href = "#3"),
                                          " ", tags$a("amino acids sequence", href = "#2_2"), " or ", tags$a("PDB codes", href = "#2_1"),
                                          " with their respective link to the ", tags$a("RCSB PDB", href = "https://www.rcsb.org/#Category-learn", target = "_blank"),
                                          " (The Protein Data Bank of the Research Collaboratory for Structural Bioinformatics)."),
                                        p("If you are interested in just one of these queries, you can skip the other in this tutorial. They are not",
                                          "depended to each other.")
                                        ),
                                    div(id = "2_1", class = "seccion level2",
                                        h3("Get PDB"),
                                        p("Generate a table with the PDB codes and their links to the ", tags$a("RCSB PDB", href = "https://www.rcsb.org/#Category-learn", target = "_blank"),
                                          " is done by clicking the \"Get PDB\" button above the ",
                                          actionLink("proteinHelp2", "Protein Table"), " as shown below."),
                                        img(src = 'Tutorial(12).jpg',
                                            width = '350px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The generated table looks like this"),
                                        img(src = 'Tutorial(13).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The columns show the EC number, the organism, the PDB code, and the link. If you click any of this link, you are going to ",
                                          "be redirected to the PDB site with an image like this"),
                                        img(src = 'Tutorial(14).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("Also, you can download the whole table clicking the blue button, or return to the Protein table section clicking the orange one")
                                        ),
                                    div(id = "2_2", class = "seccion level2",
                                        h3("Get amino acids sequence"),
                                        p("The UniProt code that some proteins have are a distinctive code provides by the ",
                                          tags$a("UniProt database,", href = "https://www.uniprot.org/help/", target = "_blank"),
                                          " that a particular enzyme has. With this code, you can access the amino acids sequence of the protein. Brenda ",
                                          "provides this code and the sequence with it. By clicking the \"Get sequence\" blue button, as shown below, a table ",
                                          "reporting the sequence found by BRENDA are going to be shown."),
                                        img(src = 'Tutorial(12).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("This table looks like this"),
                                        img(src = 'Tutorial(15).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("You can access the sequence code by download a .txt file clicking the blue button. The sequence to be downloaded can ",
                                          "be selected in the report table clicking the respective row or download all the found sequence switching the \"Download ",
                                          "all the list\" button in the sidebar.")
                                        ),
                                    div(id = "2_3", class = "seccion level2",
                                        h3("Get functional parameters"),
                                        p("The complete functional parameters query tutorial is in the ",
                                          tags$a("next section.", href = "#3"), " However, to generate a table with the 12 numerical parameters available in Brenda ",
                                          "for every protein in the Protein table, you have to activate the \"Select all functional parameters\" switch on the top ",
                                          "of sidebar, let the \"Search for all parameters\" switch activated above the table and click the \"Search for parameters\" ",
                                          "green button at the end of the sidebar."),
                                        img(src = 'Tutorial(16).jpg',
                                            width = '400px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "3", class = "seccion level1",
                                        h2("Functional Parameters Query"),
                                        p("This is the particular functionality this application has and, naturally, one of the tools that have the most options. If you", 
                                          "simply want all the numerical parameters available you can do ",
                                          tags$a("this.", href = "#2_3"))
                                        ),
                                    div(id = "3_1", class = "seccion level2",
                                        h3("Selecting parameters"),
                                        p("Brenda has 12 numerical parameters available for most of the EC number. Each of them is shown in different tables like this: "),
                                        img(src = 'Tutorial(17).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The functional parameters query generate a single table with the parameters selected separated by the organism (and the commentary) ",
                                          "in which the enzyme is located. To select these parameters, click the checkbox on the left of the list in the sidebar. If you want to ",
                                          "select them all activating the \"Select all functional parameters\" switch on the top of the sidebar."),
                                        img(src = 'Tutorial(18).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("You can also select the enzymes included in the Parameter table by clicking the rows corresponding to the organism you want. But, if the ",
                                          "\"Use all the proteins found\" switch above the table is selected, the query is going to include all the enzymes, therefore, if you want to ",
                                          "search just the selected ones, deactivate it."),
                                        img(src = 'Tutorial(19).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: Select the enzymes to the query will re-generate the Protein Table, leaving only those selected. The whole list of enzymes/organism ",
                                                 "cannot be recovered, for that, do the ",
                                                 tags$a("Enzyme query", href = "#1_2"), " again."),
                                        tags$br(),
                                        p("To do the query, click the \"Search for parameters\" green button at the end of the sidebar. This may take several minutes."),
                                        img(src = 'Tutorial(20).jpg',
                                            width = '450px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "3_2", class = "seccion level2",
                                        h3("Other options of query")
                                        ),
                                    div(id = "3_2_1", class = "seccion level3",
                                        h4("Selecting by phylogeny"),
                                        p("This functionality allows you to select the enzymes by one taxonomic rank of the organism which contains it. The taxonomic ranks available ",
                                          "are superkingdom, phylum, class, order, family, genus, and species."),
                                        img(src = 'Tutorial(73).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("To do this filter prior to the parameter query, click the \"Select by phylogeny\" yellow button. After the phylogenetic is found, above the ",
                                          "Protein table, it's going to appear a selector with the different available taxonomic ranks. Select the one you prefer, and this category is ",
                                          "going to appear as a new column on the table below."),
                                        img(src = 'Tutorial(74).jpg',
                                            width = '600px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("If you want to filter by one of these categories, let the protein table with the rows of the organism you want, and then press \"Search\" ",
                                          "button."),
                                        helpText("Note: If you select by phylogeny, the rows selected are going to have no meaning to the selection.")
                                        ),
                                    div(id = "3_2_2", class = "seccion level3",
                                        h4("Add enzymes"),
                                        p("You can add more EC numbers to the query doing ",
                                          tags$a("this.", href = "#1_2_1"))
                                        ),
                                    div(id = "3_2_3", class = "seccion level3",
                                        h4("Filter by UniProt code"),
                                        p("If you just want to use the enzymes that have ",
                                           tags$a("UniProt code,", href = "https://www.uniprot.org/help/", target = "_blank"), " activate the switch below the list of parameters."),
                                        img(src = 'Tutorial(21).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "3_3", class = "seccion level2",
                                        h3("The Parameter Table"),
                                        p("The generated table has:"),
                                        div(style = "margin-left: 40px;",
                                            p(icon("arrow-alt-circle-right"), "6 columns indicating enzyme data, same as the shown in the ",
                                              tags$a("protein table", href = "#1_2"), ":"),
                                            div(style = "margin-left: 60px;",
                                                p(icon("arrow-right"), "EC number"),
                                                p(icon("arrow-right"), "Systematic name"),
                                                p(icon("arrow-right"), "Recommended name"),
                                                p(icon("arrow-right"), "UniProt"),
                                                p(icon("arrow-right"), "(Collapsible) Commentary"),
                                                p(icon("arrow-right"), "(Collapsible) Literature: showing the Pubmed ID of the scientific paper")),
                                            p(icon("arrow-alt-circle-right"), "3 or 4 columns for each parameter, the amount of columns is determined by the dependency of the molecule, such ",
                                              "as an inhibitor or substrate:"),
                                            div(style = "margin-left: 60px;",
                                                p(icon("arrow-right"), "Value: indicating the value as a single number or as a range (independent if the parameter is defined ",
                                                  "as a range or not). The column is named with the name of the parameter, an underscore and the word \"value\""),
                                                p(icon("arrow-right"), "Substrate/Inhibitor: indicating the name of the substrate or inhibitor. If the parameter does not ",
                                                  "depend on a molecule, this columns is missing"),
                                                p(icon("arrow-right"), "(Collapsible) Commentary"),
                                                p(icon("arrow-right"), "(Collapsible) Literature"))),
                                        p("The parameter table looks like this:"),
                                        img(src = 'Tutorial(22).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("Above the table are two checkboxes indicating if the commentary and literature column is shown. If it is clicked, any of them, all the ",
                                          "commentary (or literature) columns are shown, that is, of the organism and each parameter found. If both are selected, it will look ",
                                          "like this:"),
                                        img(src = 'Tutorial(23).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "3_4", class = "seccion level2",
                                        h3("Filters"),
                                        p("With the parameters table generated, you can also filter the numerical parameters setting the minimum and/or maximum value you ",
                                          "want to preserve. For this, there are slide inputs in the sidebar of the parameters table that also have a checkbox on the left. These ",
                                          "sliders are going to be shown if the respective parameter was selected on the previous step. In case the enzyme does not have ",
                                          "any parameter reported on the Brenda database (which can be due to the lack of data reported in the literature), the slider is ",
                                          "going to report that the parameter wasn't found. In the example below, there is no IC50 reported and the Kcat/Km was not required", 
                                          " for the query."),
                                        img(src = 'Tutorial(24).jpg',
                                            width = '350px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("To use one parameter as a filter, move the slider letting the white circle at the left indicating the minimum value and the one ",
                                          "at the right indicating the maximum value your parameter is going to have. Then, click the checkbox next to the slider. Repeat this ",
                                          "for every parameter you want, and leave the checkbox unchecked if the parameter is indifferent to your filtering. Finally, click the ",
                                          "\"Filter\" blue button to generate the filtered table. In the example below, the molecular weight was used as a filter, it is shown ",
                                          "that the minimum value before and after the filtering, has changed."),
                                        img(src = 'Tutorial(25).jpg',
                                            width = '600px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        img(src = 'Tutorial(26).jpg',
                                            width = '450px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: There are a few Pubmed ID that says \"Brenda ID: XXX\" as a code. This means that searching with Brenda SOAP did not report the Pubmed ID from ",
                                                 "the BRENDA literature code used internally on Brenda. On the ", tags$a(href = "#3_5", "downloaded table,"), " the link refers to the paper on Brenda Site"),
                                        tags$br(),
                                        p("If you want to undo the filtering and recover the generated table as what it was just after the query, unchecked every checkbox on the ",
                                          "sidebar, it doesn't matter any slider input does not have the selectors on the minimum and maximum respectively. Then, click the \"",
                                          "Filter\" blue button."),
                                        img(src = 'Tutorial(27).jpg',
                                            width = '450px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "3_5", class = "seccion level2",
                                        h3("Download table"),
                                        p("The generated and filtered table can be download locally by clicking the \"Download table\" blue button above the table."),
                                        img(src = 'Tutorial(28).jpg',
                                            width = '400px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The downloaded table is a .txt file which columns are separated by the tabulator character. It contains the commentary and literature columns ",
                                          "regardless of whether they are shown or not, and it has the parameter as shown, this is, filtered."),
                                        img(src = 'Tutorial(29).jpg',
                                            width = '600px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("You can access this table as you want, for reference ",
                                          tags$a("Microsoft Office Excel", href = "https://support.office.com/en-us/article/import-or-export-text-txt-or-csv-files-5250ac4c-663c-47ce-937b-339e391393ba", target = "_blank"),
                                          "can generate a workbook from this file, remember it has a header row, it's separated by TAB character and it doesn't have the row names. With ",
                                          tags$a("R", href = "https://cran.r-project.org/", target = "_blank"), "you can open it with the following command: "),
                                        code("read.table(<table_name>, sep = \"\\t\", header = TRUE)"),
                                        tags$br(),
                                        tags$br(),
                                        p("The protein table can be downloaded too with the same format and particularities.")
                                        ),
                                    div(id = "4", class = "seccion level1",
                                        h2("Summary table and quick access"),
                                        p("This section describes 2 tools that are available at any moment in the app.")
                                        ),
                                    div(id = "4_1", class = "seccion level2",
                                        h3("Summary table and where to find it"),
                                        p("It is possible to generate a table showing how many attributes of each protein (differentiating the EC number ",
                                          "and the organism) have been found."),
                                        img(src = 'Tutorial(30).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("You can access this table in two ways. In the dashboard menu, located in the \"Protein Search\" section, or, ",
                                          "by clicking the direct access toolbar with the ", icon("table"), " icon. Once in the \"Available Information\" section ",
                                          "click the \"Refresh\" blue button to generate or update."),
                                        img(src = 'Tutorial(31).jpg',
                                            width = '350px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "4_2", class = "seccion level2",
                                        h3("Quick access toolbar"),
                                        p("At the right\\top corner, there are four buttons in the toolbar. These buttons, called quick access, allow you to ",
                                          "change the current section to:"),
                                        tags$br(),
                                        p(icon("question"), " = ", "Tutorial section"),
                                        img(src = 'Tutorial(32a).jpg',
                                            width = '100px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        tags$br(),
                                        p(icon("asterisk"), " = ", tags$a("Protein table", href = "#2")),
                                        img(src = 'Tutorial(32b).jpg',
                                            width = '100px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        tags$br(),
                                        p(icon("wrench"), " = ", tags$a("External Tools", href = "#7")),
                                        img(src = 'Tutorial(32c).jpg',
                                            width = '100px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        tags$br(),
                                        p(icon("table"), " = ", tags$a("Summary table", href = "#4_1")),
                                        img(src = 'Tutorial(32d).jpg',
                                            width = '100px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        tags$br(),
                                        p(icon("user"), " = ", tags$a("Principal page to log in or log out", href = "#1")),
                                        img(src = 'Tutorial(32e).jpg',
                                            width = '100px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "4_3", class = "seccion level2",
                                        h3("Dashboard Menu"),
                                        p("At any point in your work, you can jump to any section using the dashboard menu at the left of the site. This can be hidden ",
                                          "or show clicking the ", icon("bars"), "on the toolbar."),
                                        img(src = 'Tutorial(33).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "5", class = "seccion level1",
                                        h2("Visualize"),
                                        p("With the generated tables, there are multiple options to do. The ",
                                          tags$a("Cluster", href = "#6"), "section, ", tags$a("External tools", href = "#7"),
                                          " and this one are about that. This section shows how to use the available visualizations BEST has.")
                                        ),
                                    div(id = "5_1", class = "seccion level2",
                                        h3("Parameter found as a bar plot"),
                                        p("To start the visualization click the \"Visualize\" green button above the parameter table."),
                                        img(src = 'Tutorial(28).jpg',
                                            width = '300px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText(div(style = "display: inline-block;font-weight: bold;", "Every visualization generated will be done using the filtered ",
                                                     "parameter table, therefore, to use the no filtered one, ", tags$a("recover it.", href = "#3_4"))),
                                        tags$br(),
                                        p("The first generated graph is a bar plot showing the found data, considering the parameters selected in the ",
                                          tags$a("Functional Parameters Query", "#3"), " and the data", tags$a("filtered (or not).", href = "#3_4")),
                                        img(src = 'Tutorial(34).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: This graph is done with the ", tags$a("Plotly library.", href = "https://plot.ly/d3-js-for-r-and-shiny-charts/", target = "_blank"),
                                                 " Hence, you can hide/show the different parameters by clicking the legend on the right, save as .png, and make a ",
                                                 "different type of zooms.")
                                        ),
                                    div(id = "5_2", class = "seccion level2",
                                        h3("Distribution as boxplot"),
                                        p("Descriptive statistics quantitatively describes data. In this case, distribution can be analyzed with a boxplot that indicates ",
                                          "the quartiles, including the median. The distribution boxplot is generated by clicking the \"Watch distribution\" blue button ",
                                          "on the sidebar panel."),
                                        img(src = 'Tutorial(35).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The outliers are shown as a point indicating value, organism and the recommended name of the enzyme. Due to the different scale, the parameters ",
                                          "are shown separately. Double click in one parameter is going to hide all the other parameters, one click and that parameter are going to hide."),
                                        tags$br(),
                                        p("The parameters of the same scale are:"),
                                        div(style = "margin-left: 40px;",
                                            p(icon("arrow-right"), "Molecular Weight"),
                                            p(icon("arrow-right"), "IC50, Specific Activity"),
                                            p(icon("arrow-right"), "Kcat/Km, Turnover Number"),
                                            p(icon("arrow-right"), "Ki, Km"),
                                            p(icon("arrow-right"), "pH Optimum, pH Range, pI"),
                                            p(icon("arrow-right"), "Temperature Optimum, Temperature Range")),
                                        img(src = 'Tutorial(36).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: This graph is done with the ", tags$a("Plotly library.", href = "https://plot.ly/d3-js-for-r-and-shiny-charts/", target = "_blank"),
                                                 " Hence, you can hide/show the different parameters by clicking the legend on the right, save as .png, and make a ",
                                                 "different type of zooms."),
                                        tags$br(),
                                        tags$br(),
                                        p("To hide the outliers for further analysis or visualization you can go back to the parameter table (by going back to visualization ",
                                          "and then back to parameter table or using the ",
                                          tags$a("dashboard menu", href = "#4_3"), ") and ", tags$a("filter", href = "#3_4"), " the data to hide this datum."),
                                        img(src = 'Tutorial(37).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: this also will filter all the organism that hasn't molecular weight reported, and the data of the organism filtered.")
                                        ),
                                    div(id = "5_3", class = "seccion level2",
                                        h3("Correlation"),
                                        p("Data can be correlated, this is, two variables may have any statistical association. This association, called correlation, whether ",
                                          "casual or not. On the Parameters Found section, click the \"Watch correlation\" button, below the \"Watch distribution\" one. Two ",
                                          "green buttons will be shown."),
                                        img(src = 'Tutorial(38).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "5_3_1", class = "seccion level3",
                                        h4("Correlation as heatmap matrix"),
                                        p("A heatmap matrix represents a function of two variables as a matrix, the matrix itself represent these two dimensions and the function ",
                                          " is represented with a color code (the heatmap). In this case, the two variables are all the functional parameter pairs of the found ",
                                          "data. The heatmap represents a measure of the correlation."),
                                        tags$br(),
                                        p("Click the \"As heatmap matrix\" green button below the \"Watch correlation\" one. This is going to redirect to the generated heatmap."),
                                        img(src = 'Tutorial(39).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The correlation looks like this: "),
                                        img(src = 'Tutorial(40).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The color code (color palette) can be changed by selecting any other palette on the sidebar."),
                                        img(src = 'Tutorial(41).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "5_3_2", class = "seccion level3",
                                        h4("Correlation as a paired scatter plot"),
                                        p("The correlation also can be analyzed visually instead of numerically. To watch a paired scatter plot, click the \"As a scatterplot matrix\" ",
                                          "green button below the \"As a heatmap matrix\" one, or change the tab on the Correlation section after generating the heatmap matrix."),
                                        img(src = 'Tutorial(42).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("For storage reason, the paired scatter plot is able to have six parameters maximum, and, two minimum. Select the parameters to analyze ",
                                          "an click the \"Get correlation\" button below. Keep in mind, if the correlated observations are more than 2000, the data is going to be ",
                                          "uniformly filtered."),
                                        img(src = 'Tutorial(43).jpg',
                                            width = '250px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The generated plot showed the scatter plots relating two variables, the distribution as a density histogram and whether the data was obtained ",
                                          "from a wild type or mutant enzyme."),
                                        img(src = 'Tutorial(44).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: This graph is done with the ", tags$a("Plotly library.", href = "https://plot.ly/d3-js-for-r-and-shiny-charts/", target = "_blank"),
                                                 " Hence, you can hide/show the different parameters by clicking the legend on the right, save as .png, and make a ",
                                                 "different type of zooms.")
                                        ),
                                    div(id = "6", class = "seccion level1",
                                        h2("Cluster"),
                                        p(tags$a("Cluster analysis", href = "https://dl.acm.org/citation.cfm?doid=568574.568575", target = "_blank"),
                                          " is a technic of exploratory data mining and statistical data analysis. It consists of grouping a set of object related to each other more than ",
                                          "those on other groups (clusters). To do clustering click the \"Analyze\" green button above the parameter table."),
                                        img(src = 'Tutorial(28).jpg',
                                            width = '300px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "6_1", class = "seccion level2",
                                        h3("K-means"),
                                        p(tags$a("K-means", href = "https://ieeexplore.ieee.org/document/1056489", target = "_blank") ," is a clustering algorithm consisting of generating ",
                                          "random points and calculate a distance among the nearest point, discarding the most distant one and repeating until the k (number of clusters) are ",
                                          "generated. To do the k-means clustering click the \"K-means clustering\" green button on the Cluster section or change the tab to the K-means clustering."),
                                        img(src = 'Tutorial(45).jpg',
                                            width = '350px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("On the sidebar select 2 (for a 2-dimensional clustering) or 3 (for a 3-dimensional one) and click the \"Clusterize\" blue button below the parameters ",
                                          "list."),
                                        img(src = 'Tutorial(46).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("The k (the number of clusters) is determined by calculating the within-cluster sum of square (WSS), which a measure of the distance of the point within ",
                                          "each cluster. Then, the biggest decrease is chosen, this is called the \"Elbow\" method."),
                                        img(src = 'Tutorial(47).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: This graph is done with the ", tags$a("Plotly library.", href = "https://plot.ly/d3-js-for-r-and-shiny-charts/", target = "_blank"),
                                                 " Hence, you can hide/show the different parameters by clicking the legend on the right, save as .png, and make a ",
                                                 "different type of zooms."),
                                        tags$br(),
                                        p("You also can download the clustering as a .txt file, the format has described ", 
                                          tags$a("before,", href = "#3_5")," containing the cluster as a column with the number of the cluster. This is more useful in case the data was reduced in order to ",
                                          "allow the visualization.")
                                        ),
                                    div(id = "6_2", class = "seccion level2",
                                        h3("DBSCAN"),
                                        p("Another clustering algorithm is ", tags$a("DBSCAN,", href = "https://towardsdatascience.com/how-dbscan-works-and-why-should-i-use-it-443b4a191c80", target = "_blank"), " this algorithm ",
                                          "is based on the density of the clusters. Taking an initial epsilon, as the distance between two points, and a minimum of points value, a cluster is defined when a region of points is ",
                                          "enough dense. To do a DBSCAN clustering, click the second green button on the sidepanel of the Clustering section or just select the tab \"DBSCAN Clustering\"."),
                                        img(src = 'Tutorial(48).jpg',
                                            width = '350px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("Once in the \"DBSCAN Clustering\" section, select the two o three parameters you want to cluster and merge this data by clicking the \"Merge data\" blue button."),
                                        img(src = 'Tutorial(49).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("This going to generate a histogram showing the twenty intervals of the (Euclidean) distance between each parameter and the frequency of it. There are no rules of how to select ",
                                          "the two parameters needed, epsilon and minimum points, but the shown plot could help. Choose an adequate distance, generally the maximum value of the modal range. A too long epsilon ",
                                          "won't generate too many cluster and too small one clustering the whole data. Min Points too small or too long has the same effect. Min Points should be at least greater than the number ",
                                          "of dimension (two or three parameters) plus one. Use the sliders to select these parameters and click the \"Cluster\" button."),
                                        img(src = 'Tutorial(50).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: it is possible that the amount of data are too much to calculate distance. In this case, data is going to be reduced and the eliminated data are not going to be used for ",
                                                 "clustering."),
                                        p("Clusters are defined as numbers from 1, while 0 means unclustered points, due to low-density regions is generated, also called as outliers."),
                                        img(src = 'Tutorial(51).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: This graph is done with the ", tags$a("Plotly library.", href = "https://plot.ly/d3-js-for-r-and-shiny-charts/", target = "_blank"),
                                                 " Hence, you can hide/show the different parameters by clicking the legend on the right, save as .png, and make a ",
                                                 "different type of zooms."),
                                        p("Same as \"", tags$a("k-means", href = "#6_1"), "\", if data is lost for visualization, this is available clustering in a downloaded table.")
                                        ),
                                    div(id = "7", class = "seccion level1",
                                        h2("External tools"),
                                        p("To make all this data useful we propose two other tools available online ",
                                          tags$a("Scratch Proteomics", href = "http://scratch.proteomics.ics.uci.edu/", taget = "_blank"), " and ", tags$a("BLAST.", href = "https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome", target = "_blank"),
                                          " You can find this section by ", tags$a("dashboard menu", href = "#4_3"), " or by ", tags$a("quick access toolbar.", href = "#4_2"))
                                        ),
                                    div(id = "7_1", class = "seccion level2",
                                        h3("Protein Prediction"),
                                        p("University of California, Irvine, has developed a Protein Predictor. Using RNN (Recursive Neural Network), this tool can predict properties such as domains, secondary structure, tertiary structure ",
                                          "and more (", tags$a(href = "http://scratch.proteomics.ics.uci.edu/explanation.html", "more info", target = "_blank"), ")."),
                                        p("To help you use this tool, we allow you to copy one of your obtained sequences. First, click the blue button below the \"Scratch Protein Prediction\" box."),
                                        img(src = 'Tutorial(52).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("This is going to redirect you to ", tags$a("FASTA section,", href="#2_2"), " this has a white button which says \"Copy a Sequence to Clipboard\". Select one (and just one)",
                                          " sequence on the table and click the button."),
                                        img(src = 'Tutorial(53).jpg',
                                            width = '600px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("Note: Make sure the sequence is available (Found_using_Brenda = true) if not, nothing is going to be copied to your clipboard."),
                                        p("Now the green button is going to have a ticket, this means a sequence is in your clipboard. Press the green button and you are going to be redirected to the Scratch proteomics site."),
                                        img(src = 'Tutorial(54).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("In this page, you can select what predictor you want to use. Enter your email and in the Protein Sequence space, paste the sequence which is already on your clipboard."),
                                        img(src = 'Tutorial(55).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        helpText("The header is missing for requirements of the tool. Please, read the tool specifications to know if your sequence is allowed.")
                                        ),
                                    div(id = "7_2", class = "seccion level2",
                                        h3("Sequence Alignment"),
                                        p("Basic Local Alignment Search Tool (BLAST) is a suite of programs for aligning sequence. These programs include Standard Nucleotide BLAST, for the nucleotide sequence, and ",
                                          "Standard Protein BLAST, for amino acids sequence. This last one receives a number of inputs (",
                                          tags$a("more info", href = "https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&PAGE_TYPE=BlastDocs&DOC_TYPE=BlastHelp", target = "_blank"),
                                          "), including upload a text file with the sequence in FASTA format as the one ", tags$a("you can obtain", href = "#2_2"), " using BEST."),
                                        img(src = 'Tutorial(56).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "8", class = "seccion level1",
                                        h2("Saved tables"),
                                        p("Brenda Easy Search Tool saves your last query for 72 hours in order to do your analysis easier. If you already make a query before, there will be a message on the toolbar."),
                                        img(src = 'Tutorial(57).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("To load the saved table, click the \"more info\" link to be redirected."),
                                        img(src = 'Tutorial(58).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("On the table in the right panel, select the tables you want to load. Particularly, the saved protein table, which contains the organisms and type of enzyme according to the ",
                                          " EC number, will be going to be load whether it was selected or not."),
                                        img(src = 'Tutorial(59).jpg',
                                            width = '700px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("After the selection is done, click the \"Load Selected Table\" blue button below the table. The selected table (include the Protein table) is going to be regenerated and the ",
                                          "not selected are going to be erased. If you change your mind and don't want to save this tables anymore, change the section (using ", tags$a("quick access or dashboard menu",
                                          href = "#4"), ") or click the \"Search another enzyme\" red button on the sidebar and the tables are going to be erased."),
                                        img(src = 'Tutorial(60).jpg',
                                            width = '200px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9", class = "seccion level1",
                                        h2("Errors"),
                                        p("This is an application on development, hence, we are still dealing with a few minor issues. Some of them are identified and are presented as ",
                                          tags$a("Unavoidable errors", href = "#9_2"), " and the not identified yet can be reported on the ",
                                          actionLink("sugghelp", "Suggestions section."), " The ones that are identified, but depends on the user are presented as ",
                                          tags$a("Avoidable errors", href = "#9_1"), "on the section below.")
                                        ),
                                    div(id = "9_1", class = "seccion level2",
                                        h3("Avoidable errors"),
                                        p("If the user, on purpose or accidentally, enters an invalid input that is identified, an error is going to be announced and the action won't be done. Those kinds of error are ",
                                          "listed here.")
                                        ),
                                    div(id = "9_1_1", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Enter Brenda User"),
                                        p("Without a valid Brenda User, we cannot do any query, particularly, the first one: the Enzyme query."),
                                        img(src = 'Tutorial(61).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_1_2", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Wrong Password"),
                                        p("The entered password is invalid, but the mail is a valid Brenda User."),
                                        img(src = 'Tutorial(62).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_1_3", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Not a Brenda Account"),
                                        p("The entered mail hasn't registered on Brenda, to register, please enter on the ",
                                          tags$a("Brenda Registration site.", href = "https://www.brenda-enzymes.org/register.php", target = "_blank")),
                                        img(src = 'Tutorial(63).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_1_4", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Make an enzyme query"),
                                        p("To be able to get amino acids FASTA sequence, PDB, or functional parameters, an ",
                                          tags$a("enzyme query", href = "#1_1"), " must be done first."),
                                        img(src = 'Tutorial(64).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_1_5", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Cannot add"),
                                        p("Simply, if enzyme hasn't added to the query, there's no need to add more."),
                                        img(src = 'Tutorial(65).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_1_6", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Make a functional parameters query"),
                                        p("The visualization and the clustering analysis need functional parameters, therefore, to apply these tools, first, do a ",
                                          tags$a("functional parameters query).", href = "#3")),
                                        img(src = 'Tutorial(66).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_1_7", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "A wrong number of parameters selected"),
                                        p("Some tools require selecting some of the functional parameters found. The number of these parameters are indicated, however, if ",
                                          "the amount of selected parameter differs from those indicated, you're going to be notified and the action won't be done."),
                                        img(src = 'Tutorial(67).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_2", class = "seccion level2",
                                        h3("Unavoidable errors"),
                                        p("Some issues are identified, but we haven't found a solution yet. Some of them have a not reported solution, but some not. The objective of this section ",
                                          "is to make those issues clear and make the user be aware of them.")
                                        ),
                                    div(id = "9_2_1", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Slower Algorithm"),
                                        p("Some enzyme has a particular problem. Due to storage issues, some search needs a slower algorithm. This is because, in order to save time, the SOAP query is made ",
                                          "just one time per EC number, and then the answer is parsed. When there are too many sequences or organisms this can't be done, so the query must be done different ",
                                          "and slower. You can cancel the operation or go ahead if you have enough time."),
                                        img(src = 'Tutorial(68).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_2_2", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Unexpected error on search"),
                                        p("This error isn't supposed to happen, but if it does you can try again. If this error persists, please, report it on the Suggestion section."),
                                        img(src = 'Tutorial(69).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        p("If you click the \"Report this Bug\" button, you are going to be redirected to the ", tags$a("Suggestion section,", href = "#9_2_6"), " where you can report it ",
                                          "following the given instruction."),
                                        img(src = 'Tutorial(70).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_2_3", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "The parameters appear selected, but the query doesn't show anything"),
                                        p("We don't know what this happen, but it did happen several times on the tests. However, there is a simple solution. First, deactivate the \"Select all\" blue switch, ",
                                          "then, select the parameters you need or reactivate the blue switch. Do the query again, in case this is still happening, repeat the process or report it ", 
                                          "in the ", actionLink("sugghelp2", "Suggestions section.")),
                                        img(src = 'Tutorial(71).jpg',
                                            width = '450px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        ),
                                    div(id = "9_2_4", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Reduce the obtained rows"),
                                        p("Some visualization and analysis require too much storage. In order to be able to do them anyway and avoid the complete failure of the web site, we choose to ",
                                          "reduce data in case the generated rows are too much to storage or to merge. This could happen on the ", tags$a("clustering", href = "#6"), " or on the ",
                                          tags$a("paired scatter plot correlation.", href = "#5_3_2"), " In the first case, the clustering is done with all the data, but the visualization doesn't. In the ",
                                          "second situation, the visualization is due with reduced data."),
                                        img(src = 'Tutorial(72).jpg',
                                            width = '400px',
                                            style = "display: block; margin-left: auto; margin-right: auto;"),
                                        tags$br(),
                                        p("To reduce the data, we eliminate the pair rows and leave the odd ones. In case there's still too much data, the process is repeated.")
                                        ),
                                    div(id = "9_2_5", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Empty table"),
                                        p("If this happens it can be for two reasons: 1) enzyme doesn't have data, but it wasn't properly reported; or 2) there was an error on search. You can ",
                                          tags$a("report this,", href = "#9_2_6"), " try again, make a search about the enzyme or contact us.")
                                        ),
                                    div(id = "9_2_6", class = "seccion level3",
                                        h4(icon("exclamation-triangle"), "Suggestions"),
                                        p("If any strange or unexpected thing happens, please report it on the ", actionLink("sugghelp3", "\"Suggestions section\".")),
                                        img(src = 'Tutorial(70).jpg',
                                            width = '500px',
                                            style = "display: block; margin-left: auto; margin-right: auto;")
                                        )
                                    ))
                    ),
                    tabItem(
                      tabName = "acknowledgments",
                      box(width = 3, height = '90vh', title = "Summary", solidHeader = TRUE, status = "primary",
                          div(style='height:80vh; overflow-y: scroll',
                              tags$li(tags$a(href = "#A1", "For the interactive web site")),
                              tags$li(tags$a(href = "#A2", "For the SOAP queries logic")),
                              tags$li(tags$a(href = "#A3", "For the Data Tables")),
                              tags$li(tags$a(href = "#A4", "For the Visualization")),
                              tags$li(tags$a(href = "#A5", "For Processing of the Text")),
                              tags$li(tags$a(href = "#A6", "For calculations")),
                              tags$li(tags$a(href = "#A7", "For generating Phylogeny")),
                              tags$li(tags$a(href = "#A8", "Reference"))
                          )
                      ),
                      box(width = 9, height = '90vh', title = h1(icon("award"), "Acknowledgments"), solidHeader = TRUE, status = "primary",
                          div(style='height:70vh; overflow-y: scroll',
                              p("This application was developed as a project of ", tags$a("CeBiB", href = "https://cebib.cl/", taget = "_blank"), "(Centre for Biotechnology and Bioengineering). ",
                                "And it couldn't be done without the following packages and tools."),
                              div(id = "A1", class = "seccion level1",
                                  h2("For the interactive web site"),
                                  p("The whole implementation of the web site was done using the R language project (", tags$a("R Core Team, 2018", href = "#RCore"), ") ",
                                    "using the software for computing environment ", tags$a("R Studio.", href = "https://www.rstudio.com/", target = "_blank")),
                                  p("Fundamental for the graphical interface and website logic are the R packages Shiny (", tags$a("Chang et. al, 2018", href = "#Chang"), ") ",
                                    "and Shiny Dashboard (", tags$a("Chang & Riberio, 2018", href = "#ChangRiberio"), ")."),
                                  p("To increase the input types, do notification errors and generate help tooltip, they were used, respectively, the following R packages: ",
                                    "Shiny Widgets (", tags$a("Perrier et al., 2018", href = "#Perrier"), "), Shiny Alert (", tags$a("Attali & Edwards, 2018", href = "#AttaliEdwards"),
                                    ") and ShinyBS (", tags$a("Bailey, 2015", href = "#Bailey"), ")."),
                                  p("To complement the website functionality, such as hide/show elements, disable/enable and activate from the server was used the R package Shinyjs (",
                                    tags$a("Attali, 2018", href = "#Attali"), ").")
                              ),
                              div(id = "A2", class = "seccion level1",
                                  h2("For the SOAP queries logic"),
                                  p("To do a query through SOAP, it was developed a ", tags$a("Java Project.", href = "https://www.java.com/en/"), " This ", tags$a("Java program", href = "https://github.com/StarBrand/BrendaSOAP", target = "_blank"),
                                    " was written using ", tags$a("IntelliJ", href = "https://www.jetbrains.com/idea/", target = "_blank"), " an IDE for JVM developed by ", tags$a("Jet Brains.", href = "https://www.jetbrains.com/", target = "_blank")),
                                  p("SOAP web service was successfully implemented thanks to the ", tags$a("Apache Axis library", href = "http://axis.apache.org/axis/", target = "_blank"), "(",
                                    tags$a("Irani and Basha, 2001", href = "#Irani"), ")."),
                                  p("To communicate SOAP Java with R based web site was used the package rJava (", tags$a("Urbanek, 2018", href = "#Urbanek"), ").")
                              ),
                              div(id = "A3", class = "seccion level1",
                                  h2("For the Data Tables"),
                                  p("Data table has a better visualization and processing using DT (", tags$a("Xie et al., 2018", href = "#Xie"), ") and \"data.table\" (",
                                    tags$a("Dowle and Srinivasan, 2019", href = "#Dowle"), ")."),
                                  p("Also, to work on the tables was used rlist (", tags$a("Ren, 2016", href = "#Ren"), ") for process list and plytr (", tags$a("Wickham, 2011", href = "#Wickham"), ") for advanced table function.")
                              ),
                              div(id = "A4", class = "seccion level1",
                                  h2("For the Visualization"),
                                  p("The graphs are generated using the packages ggplot2 (", tags$a("Wickham, 2016", href = "#Wickham2"), ") and plotly (", tags$a("Sievert, 2018", href = "#Sievert"), ").")
                              ),
                              div(id = "A5", class = "seccion level1",
                                  h2("For Processing of the Text"),
                                  p("Biological text, such as aa sequence on FASTA format, is used biostrings (", tags$a("Pagès, 2019", href = "#Pages"), ") and for more string function stringr ",
                                    "(", tags$a("Wickham, 2019", href = "#Wickham3"), ") is used. To have access to the clipboard is used rclipboard (", tags$a("Bihorel, 2017", href = "#Bihorel"), ").")
                              ),
                              div(id = "A6", class = "seccion level1",
                                  h2("For calculations"),
                                  p("Clustering analyses were implemented using the base R function and the DBSCAN package (", tags$a("Hahsler and Piekenbrock, 2018", href = "#Hahsler"), ").")
                              ),
                              div(id = "A7", class = "seccion level1",
                                  h2("For generating Phylogeny"),
                                  p("This function is not implemented yet. But, at the time, packages tath were using are taxize (", tags$a("Chamberlain et al., 2019", href = "#Chamberlain"), ") and ",
                                    "ggtree (", tags$a("Yu et al., 2017", href = "#Yu"), ").")
                              ),
                              div(id = "A8", class = "seccion level1",
                                  h2("References")
                              ),
                              div(style="line-height: 1.35; margin-left: 2em; text-indent:-2em;",
                                  div(id = "Attali", class = "seccion level2",
                                      p("Attali, Dean. ", tags$i("Shinyjs: Easily Improve the User Experience of Your Shiny Apps in Seconds,"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=shinyjs", "https://CRAN.R-project.org/pacFkage=shinyjs", target = "_blank")
                                        )
                                  ),
                                  div(id = "AttaliEdwards", class = "seccion level2",
                                      p("Attali, Dean, and Tristan Edwards. ", tags$i("Shinyalert: Easily Create Pretty Popup Messages (Modals) in \'Shiny\',"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=shinyalert", "https://CRAN.R-project.org/package=shinyalert", target = "_blank")
                                      )
                                  ),
                                  div(id = "Bailey", class = "seccion level2",
                                      p("Bailey, Eric. ", tags$i("ShinyBS: Twitter Bootstrap Components for Shiny,"),
                                        " 2015. ", tags$a(href="https://CRAN.R-project.org/package=shinyBS", "https://CRAN.R-project.org/package=shinyBS", target = "_blank")
                                      )
                                  ),
                                  div(id = "Bihorel", class = "seccion level2",
                                      p("Bihorel, Sebastien. ", tags$i("Rclipboard: Shiny/R Wrapper for \'clipboard.js\',"),
                                        " 2017. ", tags$a(href="https://CRAN.R-project.org/package=rclipboard", "https://CRAN.R-project.org/package=rclipboard", target = "_blank")
                                      )
                                  ),
                                  div(id = "Chamberlain", class = "seccion level2",
                                      p("Chamberlain, Scott, Eduard Szoecs, Zachary Foster, Zebulun Arendsee, Carl Boettiger, Karthik Ram, Ignasi Bartomeus, et al. ", tags$i("Taxize: Taxonomic Information from around the Web,"),
                                        " 2019. ", tags$a(href="https://github.com/ropensci/taxize", "https://github.com/ropensci/taxize", target = "_blank")
                                      )
                                  ),
                                  div(id = "Chang", class = "seccion level2",
                                      p("Chang, Winston, Joe Cheng, J. J. Allaire, Yihui Xie, and Jonathan McPherson. ", tags$i("Shiny: Web Application Framework for R,"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=shiny", "https://CRAN.R-project.org/package=shiny", target = "_blank")
                                      )
                                  ),
                                  div(id = "ChangRiberio", class = "seccion level2",
                                      p("Chang, Winston., Riberio, Barbara Borges. ", tags$i("Shinydashboard: Create Dashboards with \'Shiny\',"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=shinydashboard", "https://CRAN.R-project.org/package=shinydashboard", target = "_blank")
                                      )
                                  ),
                                  div(id = "Dowle", class = "seccion level2",
                                      p("Dowle, Matt, and Arun Srinivasan. ", tags$i("Data.Table: Extension of \'data.Frame\',"),
                                        " 2019. ", tags$a(href="https://CRAN.R-project.org/package=data.table", "https://CRAN.R-project.org/package=data.table", target = "_blank")
                                      )
                                  ),
                                  div(id = "Hahsler", class = "seccion level2",
                                      p("Hahsler, Michael, and Matthew Piekenbrock. ", tags$i("Dbscan: Density Based Clustering of Applications with Noise (DBSCAN) and Related Algorithms,"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=dbscan", "https://CRAN.R-project.org/package=dbscan", target = "_blank")
                                      )
                                  ),
                                  div(id = "Irani", class = "seccion level2",
                                      p("Irani, Romin, and S. Jeelani Basha. ", tags$i("AXIS: The Next Generation of Java Soap."),
                                        " 1st ed. Birmingham, UK, UK: Wrox Press Ltd., 2002."
                                      )
                                  ),
                                  div(id = "Pages", class = "seccion level2",
                                      p("Pagès, H., P. Aboyoun, R. Gentleman, and S. DebRoy. ", tags$i("Biostrings: Efficient Manipulation of Biological Strings,"),
                                        " 2019."
                                      )
                                  ),
                                  div(id = "Perrier", class = "seccion level2",
                                      p("Perrier, Victor, Fanny Meyer, and David Granjon. ", tags$i("ShinyWidgets: Custom Inputs Widgets for Shiny,"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=shinyWidgets", "https://CRAN.R-project.org/package=shinyWidgets", target = "_blank")
                                      )
                                  ),
                                  div(id = "RCore", class = "seccion level2",
                                      p("R Core Team. ", tags$i("R: A Language and Environment for Statistical Computing."),
                                        " Vienna, Austria: R Foundation for Statistical Computing, 2018. ", tags$a(href="https://www.R-project.org/", "https://www.R-project.org/", target = "_blank")
                                      )
                                  ),
                                  div(id = "Ren", class = "seccion level2",
                                      p("Ren, Kun. ", tags$i("Rlist: A Toolbox for Non-Tabular Data Manipulation,"),
                                        " 2016. ", tags$a(href="https://CRAN.R-project.org/package=rlist", "https://CRAN.R-project.org/package=rlist", target = "_blank")
                                      )
                                  ),
                                  div(id = "Sievert", class = "seccion level2",
                                      p("Sievert, Carson. ", tags$i("Plotly for R,"),
                                        " 2018. ", tags$a(href="https://plotly-book.cpsievert.me", "https://plotly-book.cpsievert.me", target = "_blank")
                                      )
                                  ),
                                  div(id = "Urbanek", class = "seccion level2",
                                      p("Urbanek, Simon. ", tags$i("RJava: Low-Level R to Java Interface,"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=rJava", "https://CRAN.R-project.org/package=rJava", target = "_blank")
                                      )
                                  ),
                                  div(id = "Wickham2", class = "seccion level2",
                                      p("Wickham, Hadley. ", tags$i("Ggplot2: Elegant Graphics for Data Analysis."),
                                        " Springer-Verlag New York, 2016. ", tags$a(href="http://ggplot2.org", "http://ggplot2.org", target = "_blank")
                                      )
                                  ),
                                  div(id = "Wickham3", class = "seccion level2",
                                      p("Wickham, Hadley. ", tags$i("Stringr: Simple, Consistent Wrappers for Common String Operations,"),
                                        " 2019. ", tags$a(href="https://CRAN.R-project.org/package=stringr", "https://CRAN.R-project.org/package=stringr", target = "_blank")
                                      )
                                  ),
                                  div(id = "Wickham", class = "seccion level2",
                                      p("Wickham, Hadley. The Split-Apply-Combine Strategy for Data Analysis.", tags$i("Journal of Statistical Software"),
                                        " 40, no. 1 (2011): 1-29.")
                                  ),
                                  div(id = "Xie", class = "seccion level2",
                                      p("Xie, Yihui, Joe Cheng, and Xianying Tan. ", tags$i("DT: A Wrapper of the JavaScript Library \'DataTables\'"),
                                        " 2018. ", tags$a(href="https://CRAN.R-project.org/package=DT", "https://CRAN.R-project.org/package=DT", target = "_blank")
                                      )
                                  ),
                                  div(id = "Yu", class = "seccion level2",
                                      p("Yu, Guangchuang, David Smith, Huachen Zhu, Yi Guan, and Tommy Tsan-Yuk Lam. Ggtree: An R Package for Visualization and Annotation of Phylogenetic Trees with Their Covariates and Other Associated Data ", tags$i("Methods in Ecology and Evolution"),
                                        " 8, no. 1 (2017): 28-36. ", tags$a(href="https://doi.org/10.1111/2041-210X.12628", "https://doi.org/10.1111/2041-210X.12628", target = "_blank")
                                      )
                                  )
                              )
                          )
                      )
                    ),
                    tabItem(
                      tabName = "faq",
                      box(width = 3, height = '90vh', title = "Summary", solidHeader = TRUE, status = "primary",
                          div(style='height:80vh; overflow-y: scroll',
                              tags$li(tags$a(href = "#F1", "Why Brenda account is needed?")),
                              tags$li(tags$a(href = "#F2", "What can we do with your Brenda account?")))
                          
                      ),
                      box(width = 9, height = '90vh', title = h1(icon("question"), "Frequently Asked Questions"), solidHeader = TRUE, status = "primary",
                          div(style='height:70vh; overflow-y: scroll',
                              div(id = "F1", class = "seccion level1",
                                  h2("Why Brenda account is needed?"),
                                  p("The query is made by ",
                                    tags$a("Brenda SOAP", href = "https://www.brenda-enzymes.org/soap.php", target = "_blank"),
                                    " which requires a Brenda account. Because of the possibility of more than one query at the same time ",
                                    "we use the personal account for that.")
                              ),
                              div(id = "F2", class = "seccion level1",
                                  h2("What can we do with your Brenda account?"),
                                  p("Brenda is used just to make every query. We also define your user to separate your generated tables ",
                                    "from other tables. But, we can access your password for anything else. If you doubt it, you can see the ",
                                    "Java code that makes every query: ",
                                    tags$a("Brenda SOAP Github repository", href = "https://github.com/StarBrand/BrendaSOAP", target = "_blank"),
                                    "To make it clear, if we could access to your password and Brenda account for personal use, we would only be able ",
                                    "to request a change of password (but cannot changing without your mail) and make queries in Brenda.")
                              )
                          )
                      )
                    )
                    
                    )
                  )
  )
   
)

# Define server logic required to draw a histogram
server <- function(session, input, output) {
   
}

# Run the application 
shinyApp(ui = ui, server = server)

